### ISSUE TITLE
questions regarding the paper

### ISSUE THREAD

**Issue by fawazsammani** [aimagelab/meshed-memory-transformer#4]
Hello. Congratulations on your brilliant work!
I'd like to ask some questions regarding the paper:
In section 4.3, you mentioned that 

> we firstly introduce a reduced version of our approach in which the i-th decoder layer is only connected to the corresponding i-th encoder layer (1-to-1), instead of being connected to all encoders.

Is that step included in your code? As what i see, your query is all the same for the visual attention in the decoder part, as you are doing:

```
        enc_att1 = self.enc_att(self_att, enc_output[:, 0], enc_output[:, 0], mask_enc_att) * mask_pad
        enc_att2 = self.enc_att(self_att, enc_output[:, 1], enc_output[:, 1], mask_enc_att) * mask_pad
        enc_att3 = self.enc_att(self_att, enc_output[:, 2], enc_output[:, 2], mask_enc_att) * mask_pad
```
Do you mean that you set alphas to 1 (simply take the sum of all encoder layers)? Because if the ith decoder layer is connected to the ith encoder layer, that means the queries are different. And may i also kindly know if you have examined the importance of taking a weighted sum rather than a sum of the encoder layers?

**Comment by baraldilorenzo** [comment]
Hi @fawazsammani,

thanks for your interest in this work! The lines of code you are citing are basically implementing Eq. 6 and 8 in the paper (i.e. the cross-attention in which each decoder layer takes all encoder layer). Being this a cross-attention, queries always come from the decoder, while keys/values are from the encoder (see Eq. 7).
Regarding the 1-to-1 experiment: that's not included in the code, but could be implemented by setting `enc_att = enc_attx` (where `x` is the index of the current decoding layer) [here](https://github.com/aimagelab/meshed-memory-transformer/blob/d992129d8a08af7cff8a998035f2b791fdf9ca6e/models/transformer/decoders.py#L49) and eventually removing all unnecessary lines.
Finally, yes, we tested with a plain sum but found that it leads to worse results than using the weighted sum.

If you have other doubts do no hesitate to contact!

Best,
Lorenzo.

**Comment by fawazsammani** [comment]
Hi. Thanks for your reply. I just want to ask regarding the meshed transformer that significantly improves the result. I think it should be a default step to do in every transformer. For your meshed decoder, are there other important things to make it work? I am using harvardnlp code for transformer, i added the meshed decoder part but the performance doesn't improve at all. Are there specific hyperparameters you set to make it perform as expected (other than 3 layers for transformer and warmup with 10000? You can have a look at the changes below:

```
class DecoderLayer(nn.Module):
    def __init__(self, size, self_attn, src_attn, feed_forward, dropout):
        super(DecoderLayer, self).__init__()
        self.size = size
        self.self_attn = self_attn
        self.src_attn = src_attn
        self.feed_forward = feed_forward
        self.sublayer = clones(SublayerConnection(size, dropout), 3)
        self.fc_alpha1 = nn.Linear(d_model * 2, d_model)
        self.fc_alpha2 = nn.Linear(d_model * 2, d_model)
        self.fc_alpha3 = nn.Linear(d_model * 2, d_model)
 
    def forward(self, x, memory, src_mask, tgt_mask):
        query = self.sublayer[0](x, lambda x: self.self_attn(x, x, x, tgt_mask))
        enc_att1 = self.src_attn(query, memory[:, 0], memory[:, 0], src_mask)
        enc_att2 = self.src_attn(query, memory[:, 1], memory[:, 1], src_mask)
        enc_att3 = self.src_attn(query, memory[:, 2], memory[:, 2], src_mask)
        alpha1 = torch.sigmoid(self.fc_alpha1(torch.cat([query, enc_att1], -1)))
        alpha2 = torch.sigmoid(self.fc_alpha2(torch.cat([query, enc_att2], -1)))
        alpha3 = torch.sigmoid(self.fc_alpha3(torch.cat([query, enc_att3], -1)))
        sum_src_attn = (enc_att1 * alpha1 + enc_att2 * alpha2 + enc_att3 * alpha3) / np.sqrt(3)
        out_sum = self.sublayer[1](sum_src_attn, None)
        return self.sublayer[2](out_sum, self.feed_forward)
```
where the sublayer performs: `self.norm(x + self.dropout(sublayer(x)))`

Thanks a lot for your help and your brilliant work

**Comment by HN123-123** [comment]
Sorry to bother you， but when i want to use meshed decoder  in harvardnlp code for transformer, I found that the memoty'shape is [batch_size, num_boxes, d_model] not contain the num_layers. I want to know how do you get the num_layers in memory. 
Much thanks for your reply!!!
